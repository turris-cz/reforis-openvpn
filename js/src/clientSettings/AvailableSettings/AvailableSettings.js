/*
 * Copyright (C) 2019-2022 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";

import {
    useAPIGet,
    withErrorMessage,
    withSpinnerOnSending,
    formFieldsSize,
} from "foris";

import SettingsTable from "./SettingsTable/SettingsTable";
import API_URLs from "../../API";

export default function AvailableSettings() {
    const [getSettingsResponse, getSettings] = useAPIGet(
        API_URLs.clientSettings
    );

    useEffect(() => {
        getSettings();
    }, [getSettings]);

    return (
        <div className={formFieldsSize}>
            <h2>{_("Available Settings")}</h2>
            <p>
                {_(
                    "A new OpenVPN client instance is created for each uploaded file. Please recheck the settings file if the instance is enabled and not running in a few minutes after the setup."
                )}
            </p>
            <SettingsTableWithErrorAndSpinner
                apiState={getSettingsResponse.state}
                clientSettings={getSettingsResponse.data}
            />
        </div>
    );
}

const SettingsTableWithErrorAndSpinner = withErrorMessage(
    withSpinnerOnSending(SettingsTable)
);
