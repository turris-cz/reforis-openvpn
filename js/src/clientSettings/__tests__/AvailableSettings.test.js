/*
 * Copyright (C) 2019-2022 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import {
    render,
    fireEvent,
    wait,
    getByText,
    waitForElement,
    getByLabelText,
} from "foris/testUtils/customTestRender";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";
import mockAxios from "jest-mock-axios";

import AvailableSettings from "../AvailableSettings/AvailableSettings";

describe("<AvailableSettings />", () => {
    const singleClient = {
        id: "A1",
        enabled: true,
        running: true,
        credentials: {
            username: "user",
            password: "password",
        },
    };

    async function renderTable(clients = []) {
        const { container } = render(
            <>
                <AvailableSettings />
                <div id="modal-container" />
            </>
        );
        mockAxios.mockResponse({ data: clients });
        await waitForElement(() => getByText(container, "Available Settings"));
        return container;
    }

    function getDelete(container) {
        return getByText(container, "Delete");
    }

    it("should display spinner", () => {
        const { container } = render(<AvailableSettings />);
        expect(container).toMatchSnapshot();
    });

    it("should handle GET error", async () => {
        const { container } = render(<AvailableSettings />);
        mockJSONError();
        await waitForElement(() =>
            getByText(container, "An error occurred while fetching data.")
        );
    });

    it("should render info when there are no clients", async () => {
        const container = await renderTable();
        expect(container).toMatchSnapshot();
    });

    it("should render table of settings", async () => {
        const settings = [
            { id: "A1", enabled: true, running: true },
            { id: "A2", enabled: false, running: false },
            { id: "B1", enabled: true, running: true },
            { id: "B2", enabled: false, running: false },
        ];
        const container = await renderTable(settings);
        expect(container).toMatchSnapshot();
    });

    it("should render a modal of a particular settings after hititng 'Edit' button", async () => {
        const container = await renderTable([singleClient]);

        fireEvent.click(getByText(container, "Edit"));

        expect(container).toMatchSnapshot();
    });

    it("should handle PUT ruequest in a modal after hititng 'Save' button", async () => {
        const container = await renderTable([singleClient]);

        fireEvent.click(getByText(container, "Edit"));

        fireEvent.change(getByLabelText(container, "Username"), {
            target: { value: "new_user" },
        });

        fireEvent.change(getByLabelText(container, "Password"), {
            target: { value: "testpass" },
        });

        fireEvent.click(getByText(container, "Save"));

        expect(mockAxios.put).toHaveBeenCalledWith(
            `/reforis/openvpn/api/client-settings/${singleClient.id}`,
            {
                id: "A1",
                enabled: true,
                credentials: {
                    username: "new_user",
                    password: "testpass",
                },
            },
            expect.anything()
        );

        mockAxios.mockResponse({
            data: {
                ...singleClient,
                credentials: {
                    username: "new_user",
                    password: "testpass",
                },
            },
        });
        expect(container).toMatchSnapshot();
    });

    it("should send PUT request", async () => {
        const container = await renderTable([singleClient]);
        fireEvent.click(getByText(container, singleClient.id));
        expect(mockAxios.put).toBeCalledWith(
            `/reforis/openvpn/api/client-settings/${singleClient.id}`,
            { enabled: false },
            expect.anything()
        );
    });

    it("should handle PUT error", async () => {
        const container = await renderTable([singleClient]);
        fireEvent.click(getByText(container, singleClient.id));
        const errorMessage = "API didn't handle this well";
        mockJSONError(errorMessage);
        await wait(() => {
            expect(mockSetAlert).toHaveBeenCalledWith(errorMessage);
        });
    });

    it("should send DELETE request", async () => {
        const container = await renderTable([singleClient]);
        fireEvent.click(getDelete(container));
        expect(mockAxios.delete).toBeCalledWith(
            `/reforis/openvpn/api/client-settings/${singleClient.id}`,
            expect.anything()
        );
    });

    it("should handle DELETE error", async () => {
        const container = await renderTable([singleClient]);
        fireEvent.click(getDelete(container));
        const errorMessage = "API didn't handle this well";
        mockJSONError(errorMessage);
        await wait(() => {
            expect(mockSetAlert).toHaveBeenCalledWith(errorMessage);
        });
    });

    it("should change status of client instance after refresh", async () => {
        const container = await renderTable([singleClient]);
        expect(getByText(container, "Running")).toBeTruthy();
        fireEvent.click(getByLabelText(container, "Check status"));
        expect(mockAxios.get).toHaveBeenCalledWith(
            `/reforis/openvpn/api/client-settings/${singleClient.id}`,
            expect.anything()
        );
        mockAxios.mockResponse({ data: { ...singleClient, running: false } });
        await waitForElement(() => getByText(container, "Not running"));
    });

    it("should change status of client instance after refresh", async () => {
        const container = await renderTable([singleClient]);
        fireEvent.click(getByLabelText(container, "Check status"));
        expect(mockAxios.get).toHaveBeenCalledWith(
            `/reforis/openvpn/api/client-settings/${singleClient.id}`,
            expect.anything()
        );
        const errorMessage = "API didn't handle this well";
        mockJSONError(errorMessage);
        await waitForElement(() =>
            getByText(container, "Cannot refresh status")
        );
    });
});
