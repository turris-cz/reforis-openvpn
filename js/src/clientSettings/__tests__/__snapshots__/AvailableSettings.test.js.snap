// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvailableSettings /> should display spinner 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Available Settings
    </h2>
    <p>
      A new OpenVPN client instance is created for each uploaded file. Please recheck the settings file if the instance is enabled and not running in a few minutes after the setup.
    </p>
    <div
      class="spinner-wrapper my-3 text-center"
    >
      <div
        class="spinner-border"
        role="status"
      >
        <span
          class="sr-only"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<AvailableSettings /> should handle PUT ruequest in a modal after hititng 'Save' button 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Available Settings
    </h2>
    <p>
      A new OpenVPN client instance is created for each uploaded file. Please recheck the settings file if the instance is enabled and not running in a few minutes after the setup.
    </p>
    <div
      class="table-responsive"
    >
      <table
        class="table table-hover mb-0"
      >
        <thead
          class="thead-light"
        >
          <tr>
            <th>
              Instances
            </th>
            <th
              aria-label="Actions"
              class="text-center"
              scope="col"
            />
            <th
              aria-label="Actions"
              class="text-end"
              scope="col"
            />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="align-middle"
            >
              <div
                class="form-check form-switch mb-3 null"
              >
                <input
                  checked=""
                  class="form-check-input"
                  id="1"
                  role="switch"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="1"
                >
                  A1
                </label>
              </div>
            </td>
            <td
              class="align-middle text-center"
            >
              Running
              <button
                aria-label="Check status"
                class="btn btn btn-sm d-inline-flex justify-content-center align-items-center"
                title="Check status"
                type="button"
              >
                <span>
                  <i
                    class="fa fa-redo"
                  />
                </span>
              </button>
            </td>
            <td
              class="text-end"
            >
              <div
                aria-label="Actions"
                class="btn-group btn-group-sm mb-0"
                role="group"
              >
                <button
                  class="btn btn-primary d-inline-flex justify-content-center align-items-center"
                  type="button"
                >
                  <span>
                    <span
                      class="d-xl-none"
                    >
                      <i
                        class="fas fa-edit fa-sm"
                      />
                    </span>
                    <span
                      class="d-none d-xl-block"
                    >
                      <i
                        class="fas fa-edit fa-sm me-1"
                      />
                      Edit
                    </span>
                  </span>
                </button>
                <button
                  class="btn btn-danger d-inline-flex justify-content-center align-items-center"
                  type="button"
                >
                  <span>
                    <span
                      class="d-xl-none"
                    >
                      <i
                        class="fas fa-trash fa-sm"
                      />
                    </span>
                    <span
                      class="d-none d-xl-block"
                    >
                      <i
                        class="fas fa-trash fa-sm me-1"
                      />
                      Delete
                    </span>
                  </span>
                </button>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    id="modal-container"
  >
    <div
      class="modal fade show"
      role="dialog"
    >
      <div
        class="modal-dialog modal-dialog-centered modal-dialog-scrollable"
        role="document"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title"
            >
              Edit Client Settings
            </h5>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body"
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="2"
              >
                Username
              </label>
              <div
                class="input-group"
              >
                <input
                  class="form-control"
                  disabled=""
                  id="2"
                  type="text"
                  value="new_user"
                />
              </div>
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="3"
              >
                Password
              </label>
              <div
                class="input-group"
              >
                <input
                  autocomplete="current-password"
                  class="form-control"
                  disabled=""
                  id="3"
                  type="password"
                  value="testpass"
                />
                <button
                  class="input-group-text"
                  type="button"
                >
                  <i
                    class="fa fa-eye"
                  />
                </button>
              </div>
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-secondary d-inline-flex justify-content-center align-items-center"
              type="button"
            >
              <span>
                Cancel
              </span>
            </button>
            <button
              class="btn btn-primary d-inline-flex justify-content-center align-items-center"
              type="button"
            >
              <span>
                Save
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AvailableSettings /> should render a modal of a particular settings after hititng 'Edit' button 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Available Settings
    </h2>
    <p>
      A new OpenVPN client instance is created for each uploaded file. Please recheck the settings file if the instance is enabled and not running in a few minutes after the setup.
    </p>
    <div
      class="table-responsive"
    >
      <table
        class="table table-hover mb-0"
      >
        <thead
          class="thead-light"
        >
          <tr>
            <th>
              Instances
            </th>
            <th
              aria-label="Actions"
              class="text-center"
              scope="col"
            />
            <th
              aria-label="Actions"
              class="text-end"
              scope="col"
            />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="align-middle"
            >
              <div
                class="form-check form-switch mb-3 null"
              >
                <input
                  checked=""
                  class="form-check-input"
                  id="1"
                  role="switch"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="1"
                >
                  A1
                </label>
              </div>
            </td>
            <td
              class="align-middle text-center"
            >
              Running
              <button
                aria-label="Check status"
                class="btn btn btn-sm d-inline-flex justify-content-center align-items-center"
                title="Check status"
                type="button"
              >
                <span>
                  <i
                    class="fa fa-redo"
                  />
                </span>
              </button>
            </td>
            <td
              class="text-end"
            >
              <div
                aria-label="Actions"
                class="btn-group btn-group-sm mb-0"
                role="group"
              >
                <button
                  class="btn btn-primary d-inline-flex justify-content-center align-items-center"
                  type="button"
                >
                  <span>
                    <span
                      class="d-xl-none"
                    >
                      <i
                        class="fas fa-edit fa-sm"
                      />
                    </span>
                    <span
                      class="d-none d-xl-block"
                    >
                      <i
                        class="fas fa-edit fa-sm me-1"
                      />
                      Edit
                    </span>
                  </span>
                </button>
                <button
                  class="btn btn-danger d-inline-flex justify-content-center align-items-center"
                  type="button"
                >
                  <span>
                    <span
                      class="d-xl-none"
                    >
                      <i
                        class="fas fa-trash fa-sm"
                      />
                    </span>
                    <span
                      class="d-none d-xl-block"
                    >
                      <i
                        class="fas fa-trash fa-sm me-1"
                      />
                      Delete
                    </span>
                  </span>
                </button>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    id="modal-container"
  >
    <div
      class="modal fade show"
      role="dialog"
    >
      <div
        class="modal-dialog modal-dialog-centered modal-dialog-scrollable"
        role="document"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title"
            >
              Edit Client Settings
            </h5>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body"
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="2"
              >
                Username
              </label>
              <div
                class="input-group"
              >
                <input
                  class="form-control"
                  id="2"
                  type="text"
                  value="user"
                />
              </div>
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="3"
              >
                Password
              </label>
              <div
                class="input-group"
              >
                <input
                  autocomplete="current-password"
                  class="form-control"
                  id="3"
                  type="password"
                  value="password"
                />
                <button
                  class="input-group-text"
                  type="button"
                >
                  <i
                    class="fa fa-eye"
                  />
                </button>
              </div>
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-secondary d-inline-flex justify-content-center align-items-center"
              type="button"
            >
              <span>
                Cancel
              </span>
            </button>
            <button
              class="btn btn-primary d-inline-flex justify-content-center align-items-center"
              type="button"
            >
              <span>
                Save
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AvailableSettings /> should render info when there are no clients 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Available Settings
    </h2>
    <p>
      A new OpenVPN client instance is created for each uploaded file. Please recheck the settings file if the instance is enabled and not running in a few minutes after the setup.
    </p>
    <p
      class="text-muted text-center"
    >
      There are no settings added yet.
    </p>
  </div>
  <div
    id="modal-container"
  >
    <div
      class="modal fade"
      role="dialog"
    >
      <div
        class="modal-dialog modal-dialog-centered modal-dialog-scrollable"
        role="document"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title"
            >
              Edit Client Settings
            </h5>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body"
          >
            <div
              class="spinner-wrapper text-center"
            >
              <div
                class="spinner-border"
                role="status"
              >
                <span
                  class="sr-only"
                />
              </div>
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-secondary d-inline-flex justify-content-center align-items-center"
              type="button"
            >
              <span>
                Cancel
              </span>
            </button>
            <button
              class="btn btn-primary d-inline-flex justify-content-center align-items-center"
              disabled=""
              type="button"
            >
              <span>
                Save
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AvailableSettings /> should render table of settings 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Available Settings
    </h2>
    <p>
      A new OpenVPN client instance is created for each uploaded file. Please recheck the settings file if the instance is enabled and not running in a few minutes after the setup.
    </p>
    <div
      class="table-responsive"
    >
      <table
        class="table table-hover mb-0"
      >
        <thead
          class="thead-light"
        >
          <tr>
            <th>
              Instances
            </th>
            <th
              aria-label="Actions"
              class="text-center"
              scope="col"
            />
            <th
              aria-label="Actions"
              class="text-end"
              scope="col"
            />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="align-middle"
            >
              <div
                class="form-check form-switch mb-3 null"
              >
                <input
                  checked=""
                  class="form-check-input"
                  id="1"
                  role="switch"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="1"
                >
                  A1
                </label>
              </div>
            </td>
            <td
              class="align-middle text-center"
            >
              Running
              <button
                aria-label="Check status"
                class="btn btn btn-sm d-inline-flex justify-content-center align-items-center"
                title="Check status"
                type="button"
              >
                <span>
                  <i
                    class="fa fa-redo"
                  />
                </span>
              </button>
            </td>
            <td
              class="text-end"
            >
              <div
                aria-label="Actions"
                class="btn-group btn-group-sm mb-0"
                role="group"
              >
                <button
                  class="btn btn-primary d-inline-flex justify-content-center align-items-center"
                  type="button"
                >
                  <span>
                    <span
                      class="d-xl-none"
                    >
                      <i
                        class="fas fa-edit fa-sm"
                      />
                    </span>
                    <span
                      class="d-none d-xl-block"
                    >
                      <i
                        class="fas fa-edit fa-sm me-1"
                      />
                      Edit
                    </span>
                  </span>
                </button>
                <button
                  class="btn btn-danger d-inline-flex justify-content-center align-items-center"
                  type="button"
                >
                  <span>
                    <span
                      class="d-xl-none"
                    >
                      <i
                        class="fas fa-trash fa-sm"
                      />
                    </span>
                    <span
                      class="d-none d-xl-block"
                    >
                      <i
                        class="fas fa-trash fa-sm me-1"
                      />
                      Delete
                    </span>
                  </span>
                </button>
              </div>
            </td>
          </tr>
          <tr>
            <td
              class="align-middle"
            >
              <div
                class="form-check form-switch mb-3 null"
              >
                <input
                  class="form-check-input"
                  id="2"
                  role="switch"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="2"
                >
                  A2
                </label>
              </div>
            </td>
            <td
              class="align-middle text-center"
            />
            <td
              class="text-end"
            >
              <div
                aria-label="Actions"
                class="btn-group btn-group-sm mb-0"
                role="group"
              >
                <button
                  class="btn btn-primary d-inline-flex justify-content-center align-items-center"
                  type="button"
                >
                  <span>
                    <span
                      class="d-xl-none"
                    >
                      <i
                        class="fas fa-edit fa-sm"
                      />
                    </span>
                    <span
                      class="d-none d-xl-block"
                    >
                      <i
                        class="fas fa-edit fa-sm me-1"
                      />
                      Edit
                    </span>
                  </span>
                </button>
                <button
                  class="btn btn-danger d-inline-flex justify-content-center align-items-center"
                  type="button"
                >
                  <span>
                    <span
                      class="d-xl-none"
                    >
                      <i
                        class="fas fa-trash fa-sm"
                      />
                    </span>
                    <span
                      class="d-none d-xl-block"
                    >
                      <i
                        class="fas fa-trash fa-sm me-1"
                      />
                      Delete
                    </span>
                  </span>
                </button>
              </div>
            </td>
          </tr>
          <tr>
            <td
              class="align-middle"
            >
              <div
                class="form-check form-switch mb-3 null"
              >
                <input
                  checked=""
                  class="form-check-input"
                  id="3"
                  role="switch"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="3"
                >
                  B1
                </label>
              </div>
            </td>
            <td
              class="align-middle text-center"
            >
              Running
              <button
                aria-label="Check status"
                class="btn btn btn-sm d-inline-flex justify-content-center align-items-center"
                title="Check status"
                type="button"
              >
                <span>
                  <i
                    class="fa fa-redo"
                  />
                </span>
              </button>
            </td>
            <td
              class="text-end"
            >
              <div
                aria-label="Actions"
                class="btn-group btn-group-sm mb-0"
                role="group"
              >
                <button
                  class="btn btn-primary d-inline-flex justify-content-center align-items-center"
                  type="button"
                >
                  <span>
                    <span
                      class="d-xl-none"
                    >
                      <i
                        class="fas fa-edit fa-sm"
                      />
                    </span>
                    <span
                      class="d-none d-xl-block"
                    >
                      <i
                        class="fas fa-edit fa-sm me-1"
                      />
                      Edit
                    </span>
                  </span>
                </button>
                <button
                  class="btn btn-danger d-inline-flex justify-content-center align-items-center"
                  type="button"
                >
                  <span>
                    <span
                      class="d-xl-none"
                    >
                      <i
                        class="fas fa-trash fa-sm"
                      />
                    </span>
                    <span
                      class="d-none d-xl-block"
                    >
                      <i
                        class="fas fa-trash fa-sm me-1"
                      />
                      Delete
                    </span>
                  </span>
                </button>
              </div>
            </td>
          </tr>
          <tr>
            <td
              class="align-middle"
            >
              <div
                class="form-check form-switch mb-3 null"
              >
                <input
                  class="form-check-input"
                  id="4"
                  role="switch"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="4"
                >
                  B2
                </label>
              </div>
            </td>
            <td
              class="align-middle text-center"
            />
            <td
              class="text-end"
            >
              <div
                aria-label="Actions"
                class="btn-group btn-group-sm mb-0"
                role="group"
              >
                <button
                  class="btn btn-primary d-inline-flex justify-content-center align-items-center"
                  type="button"
                >
                  <span>
                    <span
                      class="d-xl-none"
                    >
                      <i
                        class="fas fa-edit fa-sm"
                      />
                    </span>
                    <span
                      class="d-none d-xl-block"
                    >
                      <i
                        class="fas fa-edit fa-sm me-1"
                      />
                      Edit
                    </span>
                  </span>
                </button>
                <button
                  class="btn btn-danger d-inline-flex justify-content-center align-items-center"
                  type="button"
                >
                  <span>
                    <span
                      class="d-xl-none"
                    >
                      <i
                        class="fas fa-trash fa-sm"
                      />
                    </span>
                    <span
                      class="d-none d-xl-block"
                    >
                      <i
                        class="fas fa-trash fa-sm me-1"
                      />
                      Delete
                    </span>
                  </span>
                </button>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    id="modal-container"
  >
    <div
      class="modal fade"
      role="dialog"
    >
      <div
        class="modal-dialog modal-dialog-centered modal-dialog-scrollable"
        role="document"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title"
            >
              Edit Client Settings
            </h5>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body"
          >
            <div
              class="spinner-wrapper text-center"
            >
              <div
                class="spinner-border"
                role="status"
              >
                <span
                  class="sr-only"
                />
              </div>
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-secondary d-inline-flex justify-content-center align-items-center"
              type="button"
            >
              <span>
                Cancel
              </span>
            </button>
            <button
              class="btn btn-primary d-inline-flex justify-content-center align-items-center"
              disabled=""
              type="button"
            >
              <span>
                Save
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
